### DISCUSSION TITLE
图片与正文间距问题

### DISCUSSION THREAD

**Discussion by tengxiao24** [sikouhjw/gdutthesis#50] [Q&A] [answered]
作者您好，在正文中插入图片，图片与上面的正文间距正常，但与图片下的正文出现空行，间距过大，请问如何解决。
<img width="272" alt="Snipaste_2022-03-18_10-45-40" src="https://user-images.githubusercontent.com/49975070/158927870-09b38cd5-990f-4ed0-bb6f-42f961d6e361.png">

**Comment by sikouhjw** [comment, marked as answer]
这部分间距调起来比较复杂，参考《LaTeX 如何在设置行距后让字体垂直居中？》[^1]。

解决方案：在导言区加上
```tex
\makeatletter
\patchcmd{\@addtocurcol}%
  {\vskip \intextsep}%
  {\edef\save@first@penalty{\the\lastpenalty}\unpenalty
   \ifnum \lastpenalty = \@M  % hopefully the OR penalty
     \unpenalty
   \else
     \penalty \save@first@penalty \relax % put it back
   \fi
   \ifnum\outputpenalty <-\@Mii
     \addvspace\intextsep
     \vskip\parskip
   \else
     \addvspace\intextsep
   \fi}%
  {\typeout{*** SUCCESS ***}}{\typeout{*** FAIL ***}}
\patchcmd{\@addtocurcol}%
  {\vskip\intextsep \ifnum\outputpenalty <-\@Mii \vskip -\parskip\fi}%
  {\ifnum\outputpenalty <-\@Mii
     \aftergroup\vskip\aftergroup\intextsep
     \aftergroup\nointerlineskip
   \else
     \vskip\intextsep
   \fi}%
  {\typeout{*** SUCCESS ***}}{\typeout{*** FAIL ***}}
\patchcmd{\@getpen}{\@M}{\@Mi}
  {\typeout{*** SUCCESS ***}}{\typeout{*** FAIL ***}}
\makeatother
```

其实模板也做了相应的处理
https://github.com/sikouhjw/gdutthesis/blob/54b0d45b9ffde767d8f8b7d007da4c0b15c07353/gdutthesis.cls#L1061-L1089
只是 latex3 的语法让这段代码没有成功 patch。

[^1]: https://www.zhihu.com/question/46618031

**Reply by tengxiao24** [reply]
感谢 我去试一下

**Reply by SiriusFang2020** [reply]
你好，我搜索了一下cls的文件格式，貌似都没有提到导言区这个结构，请问这个导言区到底是在哪里呢？

**Reply by sikouhjw** [reply]
> 你好，我搜索了一下cls的文件格式，貌似都没有提到导言区这个结构，请问这个导言区到底是在哪里呢？

tex 文件的 `\documentclass` 与 `\begin{document}` 之间。

**Reply by SiriusFang2020** [reply]
> > 你好，我搜索了一下cls的文件格式，貌似都没有提到导言区这个结构，请问这个导言区到底是在哪里呢？
> 
> tex 文件的 `\documentclass` 与 `\begin{document}` 之间。

你好，我测试了一下好像没有作用。
这是我放置你那段代码的位置：
![image](https://user-images.githubusercontent.com/59632817/159108263-8d810729-b049-413a-8c98-d6612edf8cd4.png)

然后这里是我编译后的效果，好像还是没解决问题
![image](https://user-images.githubusercontent.com/59632817/159108467-bcd163be-fe28-43d0-b58b-f25a83c9eaa7.png)

**Reply by tengxiao24** [reply]
可以试一下 图片前和图片后 都空一行  'Enter'键 空行  我就是这么解决的。

**Reply by SiriusFang2020** [reply]
> 可以试一下 图片前和图片后 都空一行 'Enter'键 空行 我就是这么解决的。

好的多谢，已解决该问题🙏

**Comment by still-modest** [comment]
> 可以试一下 图片前和图片后 都空一行 'Enter'键 空行 我就是这么解决的。
请问具体操作是什么样的？我按照您的操作在空行之后依旧存在原本的间距
![temp](https://github.com/sikouhjw/gdutthesis/assets/137246115/9479d439-e25c-4959-9a35-c0967dab4861)

**Reply by sikouhjw** [reply]
1. `style/free-float` 设置了什么？有没有设置为固定 `H`？
2. 现在的效果是怎么样的？哪里的间距不对？

**Reply by still-modest** [reply]
非常感谢您的及时回复，
1、style/free-float设置为true
2、固定H我没有设置，这个是在图片插入处设置吗？
3、现在效果仍然为图片与文字间存在较大间距，此外表格与文字间也也在较大间距
4、以上效果都是在导言区加上了您发的解决方案
![5](https://github.com/sikouhjw/gdutthesis/assets/137246115/2b3f1886-f27c-461e-bfae-ad0888157c8a)
![1](https://github.com/sikouhjw/gdutthesis/assets/137246115/2b6a6bf6-145c-4069-9ea5-24540824168c)

![temp](https://github.com/sikouhjw/gdutthesis/assets/137246115/596e7f06-6555-42a6-b7d2-73a7054a43e1)
![2](https://github.com/sikouhjw/gdutthesis/assets/137246115/767d44ac-3d5e-4ea2-a46e-4c340c24735e)
![3](https://github.com/sikouhjw/gdutthesis/assets/137246115/a4bed365-9aed-45da-aa41-2e9bc42de6ad)
![Uploading 4.png…]()

**Reply by sikouhjw** [reply]
https://github.com/sikouhjw/gdutthesis/blob/0f3745127922c301296a210dcdcd5de119c00dce/gdutthesis.cls#L1148-L1174
最新的模板不需要添加本 issue 中所提的代码，之前已经修复了这个问题。

更新 cls 文件试试，看看是不是模板版本较低导致的。

**Reply by still-modest** [reply]
我删除了 issue 中所提的代码，并跟新了cls文件，仍然不能解决问题，图片、表格与文字空格依旧会空一行。
但我采用了一个比较麻烦的办法，即在每个图片与表格下添加`\vspace{-0.6cm}`，这样可以解决问题。
![11](https://github.com/sikouhjw/gdutthesis/assets/137246115/1cfb149b-228e-4987-8367-f0788e6d1709)
![22](https://github.com/sikouhjw/gdutthesis/assets/137246115/2959c30d-c488-48c5-8c00-b974970d8ac5)
![33](https://github.com/sikouhjw/gdutthesis/assets/137246115/12a4fa8a-7e6f-48be-aa00-6a692ac34fbd)
![44](https://github.com/sikouhjw/gdutthesis/assets/137246115/ef7db278-1de2-49fa-a393-c2bda110effc)
